### DISCUSSION TITLE
How to: stateful switch?

### DISCUSSION THREAD

**Discussion by jlg89** [RavenSystem/esp-homekit-devices#1810] [MEPLHAA Scripts]
I want to use a Sonoff SV to start a generator. IO12 will be the switch to control generator start & shutoff. When the generator is running, the AC output will energize a feedback relay that will close the IO14 circuit on the Sonoff board. This way I can determine both the state of the switch (generator should be running, or not) and the actual state (generator is actually generating AC output, or not). Getting the states of both IOs in a useful way may only be achievable with a garage door type, but I'm hoping there's some way to do it with a switch. Any ideas?

The JSON I've tested is below. It works, sort of, but in HomeKit both IOs are treated as the main switch state.

```
{
  "c":{"l":13,"b":[{"g":0,"t":5}]},
  "a":[{
    "t":1,
    "0": { "r": [{ "g": 12, "v": 0 }] },
    "1": { "r": [{ "g": 12, "v": 1 }] },
    "b": [{ "g": 0 }],
    "g1":[{"g":14,"t":0}],
    "g0":[{"g":14,"t":1}]
  }]
}
```

**Comment by andyallcorn** [comment]
Two switches, one of which is the controllable start/stop, and the second just reports generator state (without being controllable)?

**Comment by jlg89** [comment]
That might just work. I'll give it a shot. Something like this, I guess:

```
{
  "c":{"l":13,"b":[{"g":0,"t":5}]},
  "a":[{
    "t":1,
    "0": { "r": [{ "g": 12, "v": 0 }] },
    "1": { "r": [{ "g": 12, "v": 1 }] },
    "b": [{ "g": 0 }],
  }],
  "a":[{
    "t":1,
    "g1":[{"g":14,"t":0}],
    "g0":[{"g":14,"t":1}]
  }]
}
```

The only catch I see is that I might need to use the "obstruction detected" feature of the garage door type, in order to turn off the activation switch if AC isn't detected from the generator within the timeout period.

**Reply by andyallcorn** [reply]
Bear with me, because (a) I may have misunderstood your requirements and (b) while this is valid, I haven’t tested the function and I may have overlooked something:

```{
	"c": {
		"l": 13,
		"b": [{
			"g": 0,
			"t": 5
		}]
	},
	"a": [{
			"t": 1,
			"0": {
				"r": [{
					"g": 12,
					"v": 0
				}],
				"m": [
					[4, 2]
				]
			},
			"1": {
				"r": [{
					"g": 12,
					"v": 1
				}],
				"m": [
					[4, 1]
				]
			},
			"b": [{
				"g": 0
			}]
		},
		{
			"t": 1,
			"ks": 0
		},
		{
			"t": 1,
			"h": 0,
			"0": {
				"m": [
					[2, -10001],
					[2, 0],
					[2, -10000]
				]
			},
			"1": {
				"m": [
					[2, -10001],
					[2, 1],
					[2, -10000],
					[4, 2]
				]
			},
			"f1": [{
				"g": 14,
				"t": 0
			}],
			"f0": [{
				"g": 14,
				"t": 1
			}]
		},
		{
			"t": 1,
			"h": 0,
			"i": 60,
			"0": {
				"m": [
					[1, 0]
				]
			}
		}
	]
}
```
Here’s the thinking behind it:

There are 4 switches here.

1 is visible in HomeKit and starts the relay.

2 is visible in HomeKit and solely reports the generator current state. It has no controls and spends its time disabled so it’s state cannot be changed physically or by HK.

3 is invisible. The GPIO feedback from the generator sets its state, and it in turn does an enable > turn on/off > disable cycle on switch 2. (Switch 2 can’t set its own state as it spends its time disabled to prevent accidental tampering.) Not sure I’ve got the GPIO state the right way round, but followed your example.

4 is invisible. It is to implement a timeout period (arbitrarily set to 60 seconds here). When switch 1 is turned on, so is this. The inching timer starts, and when it finishes it turns off triggering the 0 action (turn off switch 1). However, if either (a) switch 1 is deliberately turned off or (b) the generator starts, this has its _status_ set to off instead (which doesn’t trigger the action).

So, if I’ve written this correctly (dubious), you’ll have a switch in HomeKit that turns on your relay, and a switch you can’t command that shows when the generator is operating. When you turn on switch 1, if it doesn’t see the generator start within a minute it’ll turn off again.

**Reply by jlg89** [reply]
Daggum, you're good. That worked straight out of the box. Many thanks!

**Reply by andyallcorn** [reply]
Hah, surprised myself there then. You’re welcome.

**Reply by Andreadf67** [reply]
Great, you are really great. I would also like to have your support for this project I have in mind.
I need to check the valve connected to the GPIO5 of an esp8266 with HAA firmware.
The valve must be opened with the start button on GPIO0 and closed with the stop button on GPIO15.
I also have a switch on the GPIO4 for a full tank.
I would like to automate this:
when the start button is pressed the valve opens, if the tank is not full. 
If the tank is empty the valve remains open until the tank full signals GPIO4.
When the stop button is pressed the valve closes.
I tried with this Json but it doesn't work:
{"c": {"0": {"r": [{{"g": 5, "v": 0}]}, "b": [{"g": 0, "t": 5 }]}, "a": [{"0": {"r": [{"g": 5, "v": 0}]}, "1": {"r": [{"g" : 5, "v": 1}]}, "t": 20, "b": [{"g": 0, "t": 1}], "f0": [{"g": 15, "t": 0}], "f0": [{"g": 15}], "s": 0, "w": 0, "d": 0}, {"t": 5, "f0 ": [{" g ": 4," t ": 0}]," f1 ": [{" g ": 4}]," s ": 0}]}

![schema](https://user-images.githubusercontent.com/98307760/198029123-0ec2c241-e23c-4a86-8040-8b93af6b31b9.jpg)

**Reply by Andreadf67** [reply]
HI, on your solution I produced this:
{"c":{"0":{"r":[{"g":5,"v":}]}},
"a":[
{
"0":{"r":[{"g":5,"v":0}]},                            RELE' VALVE
"1":{"r":[{"g":5,"v":1}]},
"t":20,
"d":0},
{                                                            HIDDEN START
"1":{"m":[[1,0]]},
"t":1,
"h":0,
"f1":[{"g":0,"t":0}],
"f0":[{"g":0,"t":1}]
},
{                                                            HIDDEN STOP
"1":{"m":[[1,1]]},
"t":1,
"h":0,
"f1":[{"g":15,"t":0}],
"f0":[{"g":15,"t":1}]
},
{                                                           ACCESSORY TO CONTROL TANK LEVEL
"t":5,"ks":0
},{
"0":{"m":[[4,-10001],[4,0],[4,-10000]]},
"1":{"m":[[4,-10001],[4,1],[4,-10000]]},
"t":5,
"h":0,
"f1":[{"g":4,"t":0}],
"f0":[{"g":4,"t":1}]
},
{"1":{"s":[{"a":1}]}}
]}

YOUR EXAMPLE OF THE GENERATOR HAS BEEN VERY USEFUL TO ME.
NOW I HAVE DIFFICULTY TO UNDERSTAND HOW I CAN CHECK IF THE TANK IS FULL.
THE VALVE MUST START ONLY IF THE TANK IS EMPTY AND STOP IF THE TANK IS FULL. THANK YOU SO MUCH FOR YOUR HELP.

**Comment by jlg89** [comment]
OK, to clean this up and put a bow on it, with tremendous thanks to @andyallcorn for the JSON...

This is using a Sonoff SV, but the IO callouts can obviously be modified for anything that handles switch actions. The primary requirement is that your generator has a switch-driven start/stop option, or you've come up with a DIY apparatus. Make a YouTube video, become famous. Then put a relay on the generator's AC output that closes the IO14 loop when activated. And your network gear must be on a UPS, of course. :D

The JSON creates 4 switches:

1. Controls main relay (IO12) on Sonoff SV, i.e. starts the generator.
2. Status of feedback loop, i.e. indicates presence of AC power, visible but disabled in HomeKit.
3. Invisible to HomeKit, receives feedback signal on IO14 and sets switch 2 accordingly by enabling/changing state/disabling.
4. Invisible to HomeKit, if switch 1 turns on, but switch 3 doesn't come on within "i" seconds (i.e. generator didn't actually start for some reason), turn off switch 1. This prevents issues like the starter running down the battery, etc.

Once this is all in place, you should be able to say, with a satisfied smile, "Hey Siri, turn on the generator."

```
{
	"c": {"l": 13,"b": [{"g": 0,"t": 5}]},
	"a": [{
			"t": 1,
			"0": {"r": [{"g": 12,"v": 0}],"m": [[4, 2]]},
			"1": {"r": [{"g": 12,"v": 1}],"m": [[4, 1]]},
			"b": [{"g": 0}]
		  },
		  {"t": 1,"ks": 0},
		  {
			"t": 1,
			"h": 0,
			"0": {"m": [[2, -10001],[2, 0],[2, -10000]]},
			"1": {"m": [[2, -10001],[2, 1],[2, -10000],[4, 2]]},
			"f1": [{"g": 14,"t": 0}],
			"f0": [{"g": 14,"t": 1}]
		   },
		   {
			"t": 1,
			"h": 0,
			"i": 15,
			"0": {"m": [[1, 0]]}
		   }]
}
```

**Reply by xrust83** [reply]
"c": {"l": 13,"b": [{"g": 0,"t": 5}],"n":"Generator"},
"a": [{.....
The "n" parameter should be in the general part, not in the accessories
